### ISSUE TITLE
twisted.python.urlpath attributes (e.g., query, path, etc.) disconnected from underlying URL

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8072]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| posita reported|
|-|-|
|Trac ID|trac#8072|
|Type|release blocker: regression|
|Created|2015-10-27 06:12:32Z|
|Branch|https://github.com/twisted/twisted/tree/mutaburl-8072|

From 15.4:

```
>>> from twisted.python.urlpath import URLPath
>>> u = URLPath.fromString(b'http://foo.com/bar?key=val')
>>> u
URLPath(scheme='http', netloc='foo.com', path='/bar', query='key=val', fragment='')
>>> u.query = b''
>>> u
URLPath(scheme='http', netloc='foo.com', path='/bar', query='', fragment='')
>>> str(u)
'http://foo.com/bar'
```

From trunk:

```
>>> from twisted.python.urlpath import URLPath
>>> u = URLPath.fromString(b'http://foo.com/bar?key=val')
>>> u
URLPath(scheme='http', netloc='foo.com', path='/bar', query='key=val', fragment='')
>>> u.query = b''
>>> u
URLPath(scheme='http', netloc='foo.com', path='/bar', query='', fragment='')
>>> str(u)
'http://foo.com/bar?key=val'
```

Note the last line still has the old `query`. This appears to be because [https://github.com/twisted/twisted/blob/trunk/twisted/python/urlpath.py#L241 `URLPath.__str__`] now prints out the underlying `URLPath._url` attribute (whose own `query` attribute is not only disconnected from `URLPath`'s, but appears to be read-only).

<details><summary>Searchable metadata</summary>

```
trac-id__8072 8072
type__release_blocker__regression release blocker: regression
reporter__posita posita
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_mutaburl_8072 branches/mutaburl-8072
branch_author__hawkowl__glyph hawkowl, glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1445926352587675 1445926352587675
changetime__1446793608981520 1446793608981520
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set owner to @glyph|
|-|-|

OK, I was wondering if anyone would notice this :).  I guess someone did.  Mutability is a sucker's game.  I'll change these all to be @property attributes that re-parse the text and create a new URL object or something.

**Comment by twisted-trac** [comment]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|posita commented|
|-|-|

> I'll change these all to be @property attributes that re-parse the text and create a new URL object or something.

If you take that route, don't forget about the `URLPath.urltext` property in addition to the `__str__` method. However, I'm not sure that's the right approach unless existing code relies on mutability.

Mutability isn't necessary if there are sufficiently convenient ways to create new instances from attributes of existing instances. [str.replace](https://docs.python.org/2/library/string.html#string.replace) and [namedtuple._replace](https://docs.python.org/2/library/collections.html#collections.somenamedtuple._replace) are analogies that seem on point. `t.p.u.URL` seems to have [adopted this pattern](https://github.com/twisted/twisted/blob/trunk/twisted/python/_url.py#L423).

So another approach might be to make read-only properties that delegate to the underlying `URL` object (in other words, make `URLPath` an adapter of sorts). This would probably require adding a `URLPath.replace` (or similar) method that limits its accepted values to the same ones in `URLPath.__init__`, delegates to `URL.replace`, and calls `URLPath._fromURL` with the result.

Yet another option would be to do both (i.e., make writable properties that create a new underlying `URL` instance in addition to exposing a `replace` method), but create deprecation warnings for the setters?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

If you were going to change your code to call `URLPath.replace`, you should just change it to use `URL` instead.  This class only even _exists_ to work with existing code, so compatibility is important, and my changes thus far change the meaning of its mutability.  So I think we need to put back the magic properties, even though they are demonstrably the inferior way to do things.

As soon as Twisted 15.5 is out, start using URL :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

(In [46068]) Branching to mutaburl-8072.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @glyph|
|-|-|

Thanks for the patch!
Changes look good.
Just minor comments.

------------

Is it OK to parameterize and reuse the test method itself?

---------

Is it OK to use long, multi-line lambda expressions instead of creating a dedicated method?

---------

Is the change from pathList required? Why?

```
mapper = (lambda x, m=mapper: m(urlunquote(x)))
```

-------

For the fromBytes change, maybe we can add a comment to explain that this is required because Py3 has a different return value than Py2

--------

Please check my comments and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Replying to [adiroiban](#note_6):
> Thanks for the patch!
> Changes look good.
> Just minor comments.

Thanks for the review!

> Is it OK to parameterize and reuse the test method itself?

Yep!

> Is it OK to use long, multi-line lambda expressions instead of creating a dedicated method?

I think so.  I find plugging in inline callable objects to be more readable than defining new classes with lots of attributes.

> Is the change from pathList required? Why?

It's not technically required for this change, but it was causing some non-deterministic issues with twistedchecker; I will attempt to diagnose further later.

> For the fromBytes change, maybe we can add a comment to explain that this is required because Py3 has a different return value than Py2

OK, I'll add a comment.

> Please check my comments and merge.

Will do.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set status to `closed`|
|-|-|

(In [46089]) Merge mutaburl-8072: restore mutability of URLPath

Author: hawkowl, glyph

Reviewer: adiroiban

Fixes: [#8072](https://github.com/twisted/twisted/issues/8072)

This completes the change that makes URLPath based on
twisted.python.url; now you can change the attributes on URLPath to
mutate it as with its naive implementation.
